# Data cleaning

________________________________________________________________________________________________________________

**Filenames of scraped data**
- eye_products.json
- moisturizers_full.json
- treatments.json
- cleansers_full.json

__________________________________________________________________________________________________________________

## 0.0 Definitions

#### Variables
- brand
- product_name: product brand and name
- product_type: general category of product (i.e. essences, serums and treatments, moisturizers and creams, face wash and cleansers, eye creams and treatments, toners, exfoliators and peels)
- num_likes: number of likes
- ingredients: list of ingredients as published on Sephora
- rating: rating from 1-5 stars
- num_reviews: number of reviews received
- sensitive_type: whether product is suitable for sensitive skin
- combination_type: whether product is suitable for combination skin
- oily_type: whether product is suitable for oily skin
- normal_type: whether product is suitable for normal skin
- dry_type: whether product is suitable for dry skin
- clean_sephora: product does not contain the sulfates SLS and SLES, parabens, formaldehydes and formaldehyde-releasing agents, phthalates, mineral oil, retinylpalmitate, oxybenzone, coal tar, hydroquinone, triclosan, and triclocarban
- cruelty_free: not tested on animals
- vegan
- skin_concerns: skin care concern targeted, as highlighted on the Sephora website
- excl_ingr: specific ingredients not used in the formulation of the product
- best_for_skintype: whether it is highlighted to be best for a specific skin type
- acids: notes if the product contains any of Hyaluronic Acid, Salicylic Acid, AHA/Glycolic Acid, Vitaminc C
- award: notes if the product is an allure winner or a community favorite
- pricepervol: price per oz
- highlighted_ingr: highlighted ingredients as noted in the 'about the product' section in Sephora
- clinical_results: whether clinical results have been noted in the 'about the product' section in Sephora
- formulation_type: whether it is a cream, serum, liquid, gel, oil, lotion or other
- richness: indicates whether the product is lightweight, normal or heavy


#### Custom functions
- search_in_list --> Returns 1 if any item on the list contains the term and 0 otherwise (term - lowercase)
- extract_info2 --> Function can 1) search for term/phrase in the list and extracts the words immediately after the given phrase 2) Search for term/phrase and if found, this is added to the retrieved list 
- find_and_encode --> Takes in a dictionary containing term to be searched and the string to be returned if term is found in the list
- return_match --> Like a re.search function but fetches the exact match; none if no match 
- find_in_list --> Searches for pattern in the list and returns a list of the matches
- extract_next --> Searches for phrase in the list and extracts the next item on the list
- find_and_encode2 --> Takes in a dictionary containing the terms to be searched and the value to return if term is found; if not found, the default is returned 

__________________________________________________________________________________________________________________

## 1.0 Loading the files

In [1]:
#Import necessary libraries
import json 
import re
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from pandas_profiling import ProfileReport

In [2]:
#Adjusting settings
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)

#Allow multiple execution of commands in one cell 
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

<font color='red'> QUESTION BELOW </font>

#Building automated - NOT DEFINING THE ATTRIBUTES!
files= ['cleansers_full.json', 'eye_products.json', 'moisturizers_full.json']
raw_df=['cleansers_raw', 'eyeproducts_raw', 'moisturizers_raw']

def json_to_df(file_list, name):
    '''
    Accepts a list of files and a list of variables where the function will store the dictionary in
    '''
    for filename, variable in zip(file_list, name):
        print(filename, variable)
        with open(filename, 'r') as file:
            raw_data= json.load(file)
        variable=pd.DataFrame.from_dict(raw_data)
    
json_to_df(files, raw_df)

In [3]:
#Storing each raw json file in a dataframe 
with open('../raw_data/cleansers_full.json', 'r') as file:
    raw_data= json.load(file)
cleansers_raw=pd.DataFrame.from_dict(raw_data)

with open('../raw_data/eye_products.json', 'r') as file:
    raw_data= json.load(file)
eyeproducts_raw=pd.DataFrame.from_dict(raw_data)

with open('../raw_data/moisturizers_full.json', 'r') as file:
    raw_data= json.load(file)
moisturizers_raw=pd.DataFrame.from_dict(raw_data)

with open('../raw_data/treatments_full.json', 'r') as file:
    raw_data= json.load(file)
treatments_raw=pd.DataFrame.from_dict(raw_data)

In [4]:
#Inspecting the data
cleansers_raw.sample(5)
moisturizers_raw.sample(5)
eyeproducts_raw.sample(5)
treatments_raw.sample(5)

brand                                   name  \
402  Dr. Barbara Sturm                               Cleanser   
278        Dermalogica          Daily Microfoliant Exfoliator   
169          Indie Lee                   Brightening Cleanser   
198     Drunk Elephant  Baby Pekee Bar™ + Juju Bar Travel Duo   
38       Mario Badescu                       Buffering Lotion   

                                     about the product  \
402  [What it is: ,  \nA gentle foaming cleanser th...   
278  [What it is:  , A gentle, rice-based powder ex...   
169  [What it is: ,  A cleanser, makeup remover, an...   
198  [Which skin type is it good for?, ✔ Normal, ✔ ...   
38   [What it is: , A niacimamide and zinc oxide so...   

                                               weblink           sub_category  \
402  https://www.sephora.com/product/coconut-melt-P...  Face Wash & Cleansers   
278  https://www.sephora.com/product/watermelon-glo...            Exfoliators   
169  https://www.sephora.com/product/the-method-pol...  Face Wash & Cleansers   
198  https://www.sephora.com/product/problem-soluti...      Value & Gift Sets   
38   https://www.sephora.com/product/bi-facil-face-...                 Toners   

    main_category num_likes  \
402     cleansers        4K   
278     cleansers     48.4K   
169     cleansers     31.8K   
198     cleansers     51.2K   
38      cleansers        4K   

                                              img_link   price  \
402  https://www.sephora.com/productimages/sku/s226...  $70.00   
278  https://www.sephora.com/productimages/sku/s200...  $59.00   
169  https://www.sephora.com/productimages/sku/s207...  $34.00   
198  https://www.sephora.com/productimages/sku/s180...  $16.00   
38   https://www.sephora.com/productimages/sku/s218...  $19.00   

               size                                        ingredients  \
402    5 oz/ 150 mL  [ -Mild Tensides: Provide thorough, yet gentle...   
278     2.6 oz/ 74g  [ -Salicylic Acid: Clears clogged pores, draws...   
169  4.2 oz/ 125 mL  [ -Strawberry Seed Oil: Provides vitamins A, B...   
198            None  [ -Heilmoor Clay: Tones and detoxifies the ski...   
38      1 oz/ 29 mL  [Isopropyl Alcohol, Propylene Glycol, Zinc Oxi...   

        rating num_reviews                                         highlights  
402  4.5 stars          51  [Fragrance Free, Cruelty-Free, Good for: Acne/...  
278  4.5 stars         954  [Best for Oily, Combo, Normal Skin, Good for: ...  
169  4.5 stars         462                          [Clean at Sephora, Vegan]  
198  4.5 stars         571  [Clean at Sephora, Best for Dry, Combo, Normal...  
38     4 stars          35                                                 []

brand                                               name  \
122  OLEHENRIKSEN            Goodnight Glow Retin-ALT Sleeping Crème   
616    Lord Jones  Acid Mantle Repair Moisturizer With 250mg CBD ...   
521         COOLA  Full Spectrum 360°  Mineral Sun Silk Moisturiz...   
326   Tata Harper            Nourishing Makeup Removing Oil Cleanser   
89        Origins   Original Skin™ Matte Moisturizer with Willowherb   

                                     about the product  \
122  [Which skin type is it good for?, ✔ Normal, ✔ ...   
616  [What it is: ,  A soothing moisturizer packed ...   
521  [What it is: ,  An ultra-rich priming cream th...   
326  [What it is: , A lightweight multi-vitamin cle...   
89   [What it is: ,  A deliciously hydrating lush w...   

                                               weblink           sub_category  \
122  https://www.sephora.com/product/goodnight-glow...           Night Creams   
616  https://www.sephora.com/product/epic-moisture-...           Moisturizers   
521  https://www.sephora.com/product/birds-nest-blu...         Face Sunscreen   
326  https://www.sephora.com/product/shiseido-vital...  Face Wash & Cleansers   
89   https://www.sephora.com/product/wrinkle-warrio...           Moisturizers   

    main_category num_likes  \
122  moisturizers     50.8K   
616  moisturizers      6.3K   
521  moisturizers       997   
326  moisturizers      7.9K   
89   moisturizers     53.2K   

                                              img_link   price  \
122  https://www.sephora.com/productimages/sku/s210...  $55.00   
616  https://www.sephora.com/productimages/sku/s234...  $75.00   
521  https://www.sephora.com/productimages/sku/s233...  $42.00   
326  https://www.sephora.com/productimages/sku/s167...  $82.00   
89   https://www.sephora.com/productimages/sku/s196...  $36.00   

               size                                        ingredients  \
122   1.7 oz/ 50 mL  [ -Bakuchiol: Derived from the Ayurvedic babch...   
616   1.7 oz/ 50 mL  [-Whole Hemp Plant-Derived CBD: A restorative,...   
521   1.5 oz/ 44 mL  [-Zinc Oxide 13.5%: Provides broad-spectrum UV...   
326  4.1 oz/ 125 mL  [-Alfalfa: Nourishes skin with biotin and vita...   
89    1.7 oz/ 50 mL  [-Willowherb: Helps skin appear vibrant and sm...   

        rating num_reviews                                         highlights  
122    4 stars         478  [Clean at Sephora, AHA/Glycolic Acid, Good for...  
616    5 stars          82  [Best for Dry, Combo, Normal Skin, Good for: D...  
521  3.5 stars          15                            [Clean at Sephora, SPF]  
326    4 stars          88  [Best for Dry, Combo, Normal Skin, Good for: D...  
89   4.5 stars        2.1K             [Clean at Sephora, Community Favorite]

brand                                         name  \
2    First Aid Beauty   Eye Duty Niacinamide Brightening Eye Cream   
212  First Aid Beauty        Eye Duty Triple Remedy A.M. Gel Cream   
84           CLINIQUE  All About Eyes Serum De-Puffing Eye Massage   
98              fresh                  Black Tea Firming Eye Serum   
117             Murad              Retinol Youth Renewal Eye Serum   

                                     about the product  \
2    [What it is: ,  An illuminating eye cream that...   
212  [What it is: ,  A fast-absorbing treatment tha...   
84   [What it is:, A rollerball that massages away ...   
98   [What it is: , A fluid gel that provides a cor...   
117  [What it is:  , A visibly preventative and cor...   

                                               weblink  \
2    https://www.sephora.com/product/the-pearl-tint...   
212  https://www.sephora.com/product/spectralite-ey...   
84   https://www.sephora.com/product/sea-pack-your-...   
98   https://www.sephora.com/product/wide-awake-bri...   
117  https://www.sephora.com/product/laneige-water-...   

                sub_category main_category num_likes  \
2    Eye Creams & Treatments  eye products     36.1K   
212  Eye Creams & Treatments  eye products     23.5K   
84   Eye Creams & Treatments  eye products     20.2K   
98   Eye Creams & Treatments  eye products     11.9K   
117  Eye Creams & Treatments  eye products       15K   

                                              img_link   price           size  \
2    https://www.sephora.com/productimages/sku/s233...  $36.00  0.5 oz/ 15 mL   
212  https://www.sephora.com/productimages/sku/s178...  $36.00  0.5 oz/ 15 mL   
84   https://www.sephora.com/productimages/sku/s123...  $35.00  0.5 oz/ 15 mL   
98   https://www.sephora.com/productimages/sku/s217...  $72.00  0.5 oz/ 15 mL   
117  https://www.sephora.com/productimages/sku/s231...  $85.00  0.5 oz/ 15 mL   

                                           ingredients     rating num_reviews  \
2    [ -Niacinamide: Fades the appearance of brown ...    4 stars         283   
212  [ -Peptides: Smooth the look of skin and minim...    4 stars         200   
84   [Water, Butylene Glycol, Glycerin, Caffeine, P...    4 stars         893   
98   [ Water, Glycerin, Butylene Glycol, Pentylene ...    4 stars          56   
117  [-Retinol Tri-Active Technology: Visibly fight...  4.5 stars         155   

                                            highlights  
2    [Fragrance Free, Cruelty-Free, Best for Normal...  
212  [Clean at Sephora, Good for: Dark Circles, Goo...  
84                                                  []  
98      [Clean at Sephora, Good for: Loss of firmness]  
117  [Retinol, Good for: Anti-Aging, Best for Dry, ...

brand                                               name  \
291         Caudalie  Vinopure Natural Salicylic Acid Pore Minimizin...   
382          Primera                            Mild Facial Peeling Gel   
423  Kate Somerville           EradiKate® Daily Cleanser Acne Treatment   
147         CLINIQUE    Acne Solutions™ Cleansing Bar For Face and Body   
71   LANCER Skincare  Active Rejuvenation Serum with Triple Dermal C...   

                                     about the product  \
291  [What it is: ,  A pore-diminishing gel serum p...   
382  [What it is:  , A non-abrasive gel that improv...   
423  [What it is: ,  A medicated daily foaming acne...   
147  [What it is:  , A mild soap powered by salicyl...   
71   [What it is:  , A powerful serum that addresse...   

                                               weblink  \
291  https://www.sephora.com/product/paula-s-choice...   
382  https://www.sephora.com/product/radiance-resur...   
423  https://www.sephora.com/product/ferulic-retino...   
147  https://www.sephora.com/product/sos-save-our-s...   
71   https://www.sephora.com/product/perfectionist-...   

                  sub_category main_category num_likes  \
291                Face Serums    treatments     72.4K   
382               Facial Peels    treatments      3.4K   
423  Blemish & Acne Treatments    treatments     33.2K   
147      Face Wash & Cleansers    treatments     15.5K   
71                 Face Serums    treatments       828   

                                              img_link    price  \
291  https://www.sephora.com/productimages/sku/s211...   $49.00   
382  https://www.sephora.com/productimages/sku/s226...   $38.00   
423  https://www.sephora.com/productimages/sku/s190...   $40.00   
147  https://www.sephora.com/productimages/sku/s102...   $16.50   
71   https://www.sephora.com/productimages/sku/s237...  $150.00   

               size                                        ingredients  \
291     1 oz/ 30 mL  [  -Natural Salicylic Acid (Derived from Winte...   
382  5.0 oz/ 150 mL  [ -Broccoli Sprout Extract: Clarifies skin’s a...   
423    4 oz/ 120 mL  [ -Sulfur: Works to eliminate existing imperfe...   
147   5.2 oz/ 147 g  [Sodium Palmate, Sodium Cocoate, Water, Palmit...   
71      1 oz/ 30 mL  [ -Cell-Renewal Complex: Helps energize the lo...   

        rating num_reviews                                         highlights  
291  4.5 stars        1.1K  [Clean at Sephora, Good for: Acne/Blemishes, G...  
382    4 stars          29  [Best for Dry, Combo, Normal Skin, Good for: D...  
423  4.5 stars         572  [Good for: Acne/Blemishes, Good for: Pores , B...  
147  4.5 stars         447                                                 []  
71   4.5 stars         105                                                 []

## 2.0 Pre-processing each column

### Summary of steps taken:

1) Created new column to include both brand and product name

2) Inspected for missing values and deleted rows where appropriate

3) Removed 'stars' from rating and converted column into numeric type

4) Remove currency sign from price and converted into numeric type

5) Translated both number of likes and number of reviews into digits and converted into numeric data 

6) Created new features based on skin type the product was compatible with. This was derived from the 'about the product section'. (New columns created containing 1 or 0 for each of normal, combination, oily, dry, sensitive)

7) Created new feature based on highlights 

8) Create a new (uniform) feature for 'size'

9) Derive a feature for price adjusted for the size or volume of the product 

10) Create new feature based on highlighted ingredients

11) Create a new feature capturing whether clinical results are published in the about section

12) Extract formulation type for each ingredient and capture this in a separate feature(s)

**1) Created new column to include both brand and product name**

In [53]:
#Creating new column for full product name 
cleansers_raw['product_name']=cleansers_raw['brand']+cleansers_raw['name']
treatments_raw['product_name']=treatments_raw['brand']+treatments_raw['name']
moisturizers_raw['product_name']=moisturizers_raw['brand']+moisturizers_raw['name']
eyeproducts_raw['product_name']=eyeproducts_raw['brand']+eyeproducts_raw['name']

**2) Inspected for missing values and deleted rows where appropriate**

In [54]:
#Inspect missing values 

#see total number of items
cleansers_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 409 entries, 0 to 408
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   brand              407 non-null    object
 1   name               407 non-null    object
 2   about the product  409 non-null    object
 3   weblink            409 non-null    object
 4   sub_category       407 non-null    object
 5   main_category      409 non-null    object
 6   num_likes          407 non-null    object
 7   img_link           409 non-null    object
 8   price              403 non-null    object
 9   size               363 non-null    object
 10  ingredients        409 non-null    object
 11  rating             409 non-null    object
 12  num_reviews        402 non-null    object
 13  highlights         409 non-null    object
 14  product_name       407 non-null    object
dtypes: object(15)
memory usage: 48.1+ KB


In [55]:
cleansers_raw[cleansers_raw['size'].isnull()]

brand  \
1                Drunk Elephant   
6            SEPHORA COLLECTION   
12                       Tatcha   
36                 Estée Lauder   
48                      Clarins   
57               Drunk Elephant   
68           SEPHORA COLLECTION   
74                  MILK MAKEUP   
77                      Clarins   
80                     Shiseido   
87                        Murad   
96           REN Clean Skincare   
100             LANCER Skincare   
101                    Shiseido   
103          SEPHORA COLLECTION   
104                Estée Lauder   
105          SEPHORA COLLECTION   
137          SEPHORA COLLECTION   
138                    CLINIQUE   
139                 Dermalogica   
157          SEPHORA COLLECTION   
158                     Lancôme   
167  The Original MakeUp Eraser   
188          SEPHORA COLLECTION   
194          REN Clean Skincare   
198              Drunk Elephant   
221          SEPHORA COLLECTION   
226                       fresh   
229          SEPHORA COLLECTION   
231                   Biossance   
262  The Original MakeUp Eraser   
263                    Caudalie   
283                       SK-II   
298               Mario Badescu   
303  The Original MakeUp Eraser   
311                    CLINIQUE   
312                   trèStiQue   
323                 Bobbi Brown   
327                        None   
328                        None   
334                 MILK MAKEUP   
336     FENTY BEAUTY by Rihanna   
345             LANCER Skincare   
351                 Dermalogica   
354             LANCER Skincare   
393          SEPHORA COLLECTION   

                                                  name  \
1               Slaai™  Makeup-Melting Butter Cleanser   
6                        Cleansing & Exfoliating Wipes   
12               The Rice Polish Foaming Enzyme Powder   
36   Nutritious Super-Pomegranate Radiant Energy Lo...   
48                          Instant Eye Makeup Remover   
57                       Baby Bar Travel Duo with Case   
68                                Hemp Cleansing Wipes   
74                                     Matcha Cleanser   
77                   Hydrating Gentle Foaming Cleanser   
80                  Instant Eye And Lip Makeup Remover   
87                                     Hydrating Toner   
96                               Micro Polish Cleanser   
100          Method Intro Kit: Normal-Combination Skin   
101                         Oil Control Blotting Paper   
103         Cleansing Wipes - Aloe Vera - Moisturizing   
104  Nutritious Super-Pomegranate Radiant Energy 2-...   
105                            All-Over Solid Cleanser   
137     Exfoliating Wipes - Pink Grapefrui - Purifying   
138                Redness Solutions Soothing Cleanser   
139  Daily Microfoliant Exfoliator Full Size + Refi...   
157             Exfoliating Wipes - Bamboo - Matifying   
158                           Bi-Facil Home & Away Duo   
167   The Original MakeUp Eraser® Makeup Remover Cloth   
188         Clean Cleansing & Gentle Exfoliating Wipes   
194                   Evercalm™ Gentle Cleansing Milk    
198              Baby Pekee Bar™ + Juju Bar Travel Duo   
221           Cleansing Wipes - Watermelon - Hydrating   
226              Seaberry Skin Nutrition Cleansing Oil   
229                    Cleansing Wipes - Coconut Water   
231                   Squalane + Glycolic Renewal Mask   
262                Ride or Dye 7-Day MakeUp Eraser Set   
263                                      Beauty Elixir   
283          Facial Treatment Essence (Pitera Essence)   
298                      Witch Hazel & Rosewater Toner   
303                   Cactus 3-Piece MakeUp Eraser Set   
311  All About Clean All-in-One Cleansing Micellar ...   
312         Magic Mattifying Balm & Blotting Sheet Duo   
323                           Makeup Melter & Cleanser   
327                                               None   
328                                               None   
33

In [56]:
treatments_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 578 entries, 0 to 577
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   brand              578 non-null    object
 1   name               578 non-null    object
 2   about the product  578 non-null    object
 3   weblink            578 non-null    object
 4   sub_category       578 non-null    object
 5   main_category      578 non-null    object
 6   num_likes          578 non-null    object
 7   img_link           578 non-null    object
 8   price              573 non-null    object
 9   size               498 non-null    object
 10  ingredients        578 non-null    object
 11  rating             578 non-null    object
 12  num_reviews        572 non-null    object
 13  highlights         578 non-null    object
 14  product_name       578 non-null    object
dtypes: object(15)
memory usage: 67.9+ KB


In [57]:
treatments_raw[treatments_raw['size'].isnull()]

brand  \
13                       fresh   
14            First Aid Beauty   
18                    CLINIQUE   
19                Perricone MD   
35                       Murad   
49          SEPHORA COLLECTION   
56                  Jack Black   
57                   Dr. Jart+   
74               HUM Nutrition   
84          Kiehl's Since 1851   
88                    CLINIQUE   
90          SEPHORA COLLECTION   
92          SEPHORA COLLECTION   
99                     Clarins   
105                   Algenist   
108         SEPHORA COLLECTION   
111               IT Cosmetics   
112         REN Clean Skincare   
114                      Murad   
116                       Dior   
122                 philosophy   
129         SEPHORA COLLECTION   
131                  Dr. Jart+   
134              HUM Nutrition   
142          Peter Thomas Roth   
156                       goop   
186          Dr. Barbara Sturm   
187                  Dr. Jart+   
191                   Proactiv   
201                    Briogeo   
204                  Peace Out   
206  Dr. Dennis Gross Skincare   
213                   Proactiv   
218                  Sulwhasoo   
228               Estée Lauder   
246           First Aid Beauty   
252                 philosophy   
266                      FOREO   
272                       Dior   
280                   Shiseido   
284             Drunk Elephant   
298                  Peace Out   
324         Kiehl's Since 1851   
335             Drunk Elephant   
339                  Biossance   
348                      SK-II   
362           First Aid Beauty   
365                       Dior   
396                       Dior   
400               Estée Lauder   
405               Perricone MD   
411            BECCA Cosmetics   
413          Charlotte Tilbury   
419          Peter Thomas Roth   
424          Peter Thomas Roth   
440                  Biossance   
442                 LightStim    
449             Drunk Elephant   
453              HUM Nutrition   
457                  Dr. Jart+   
466          Peter Thomas Roth   
470                  Dr. Jart+   
500                   CLINIQUE   
503               Perricone MD   
508                  Dr. Jart+   
510                   CLINIQUE   
511          Dr. Barbara Sturm   
512         SEPHORA COLLECTION   
515                Dermalogica   
516                   CLINIQUE   
519                  Dr. Jart+   
525                   CLINIQUE   
530                 Supergoop!   
533         SEPHORA COLLECTION   
549                   CLINIQUE   
551                   CLINIQUE   
557                   Proactiv   
558         REN Clean Skincare   
561       Dr. Zenovia Skincare   
574            Kate Somerville   

                                                  name  \
13               Umbrian Clay Pore Purifying Face Mask   
14                       Ultra Repair® Hydrating Serum   
18          Clean Skin, Fresh Start Acne Solutions Set   
19                   Vitamin C Ester Brightening Serum   
35                             Rapid Collagen Infusion   
49                                   Clarifying Lotion   
56              Power Peel Multi-Acid Resurfacing Pads   
57              Focuspot™ Dark Circle Micro Tip™ Patch   
74             Big Chill™ Stress Management Supplement   
84   Clearly Corrective Accelerated Clarity & Renew...   
88   Clinique iD™:  Moisturizer + Concentrate for F...   
90                                          Face Serum   
92                                    3-in-1 Extractor   
99                                V Shaping Lift Serum   
105         REVEAL Concentrated Color Correcting Drops   
108                     Double-Ended Blemish Extractor   
111         Bye Bye Breakout Full-Coverage Concealer\t   
112                        Perfect Canvas Clean Primer   
114           Pure Skin® Clarifying Dietary Supplement   
116  Capture Youth Intense Rescue Age-Delay Revital...   
122                                            Help 

In [58]:
eyeproducts_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 220 entries, 0 to 219
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   brand              220 non-null    object
 1   name               220 non-null    object
 2   about the product  220 non-null    object
 3   weblink            220 non-null    object
 4   sub_category       220 non-null    object
 5   main_category      220 non-null    object
 6   num_likes          220 non-null    object
 7   img_link           220 non-null    object
 8   price              216 non-null    object
 9   size               190 non-null    object
 10  ingredients        220 non-null    object
 11  rating             220 non-null    object
 12  num_reviews        216 non-null    object
 13  highlights         220 non-null    object
 14  product_name       220 non-null    object
dtypes: object(15)
memory usage: 25.9+ KB


In [59]:
eyeproducts_raw[eyeproducts_raw['size'].isnull()]

brand  \
0                    Peace Out   
3                Wander Beauty   
6                       Tatcha   
8                 IT Cosmetics   
21                  KNC Beauty   
25                     Anthony   
26                Perricone MD   
33          REN Clean Skincare   
37                     Clarins   
38                   BeautyBio   
40                 INC.redible   
49          REN Clean Skincare   
52                      Kopari   
91                Perricone MD   
106                    Lancôme   
111              Wander Beauty   
120                Bobbi Brown   
126         SEPHORA COLLECTION   
127                  Biossance   
128         SEPHORA COLLECTION   
140                   Algenist   
147                      belif   
149                   Shiseido   
154               OLEHENRIKSEN   
175         SEPHORA COLLECTION   
176                      FOREO   
197                INC.redible   
202              HUM Nutrition   
218  Dr. Dennis Gross Skincare   
219                   Shiseido   

                                                  name  \
0                                    Retinol Eye Stick   
3                              Baggage Claim Eye Masks   
6          The Pearl Tinted Eye Illuminating Treatment   
8         Your Hydrating Moisturizer and Eye Cream Set   
21        All Natural Retinol Infused Eye Mask- 5 Pack   
25      High Performance Continuous Moisture Eye Cream   
26           No Makeup Concealer Broad Spectrum SPF 20   
33                     Vita Mineral™ Active 7 Eye Gel    
37                                     Eye Contour Gel   
38               GloPRO® EYE MicroTip™ Attachment Head   
40   No Puff Zone Hydrating Hemp Hydrogel Under Eye...   
49       Keep Young and Beautiful Anti-Aging Eye Cream   
52                                     Starry Eye Balm   
91                    Hypoallergenic Firming Eye Cream   
106     Advanced Génifique Yeux Light-Pearl™ Eye Serum   
111                  Baggage Claim Rose Gold Eye Masks   
120                          Vitamin Enriched Eye Base   
126                                     Clean Eye Mask   
127                         Squalane + Peptide Eye Gel   
128                                           Eye Mask   
140  Triple Algae Eye Renewal Balm with Multi-Pepti...   
147                              Moisturizing Eye Bomb   
149             Benefiance Wrinkle Smoothing Eye Cream   
154              Wrinkle Blur™ Bakuchiol Eye Gel Crème   
175                              Rose Quartz  Eye Mask   
176                    IRIS™ Illuminating Eye Massager   
197                   De-Puffing Under Eye Jade Roller   
202    Wing Man Liver Detox and Dark Circle Supplement   
218                       DRx SpectraLite™ EyeCare Pro   
219           White Lucent Anti-Dark Circles Eye Cream   

                                     about the product  \
0    [What it is:  , A concentrated serum balm that...   
3    [Which skin type is it good for?, ✔ Normal, ✔ ...   
6    [Which skin type is it good for?, ✔ Normal, ✔ ...   
8    [What it is:  , An essential duo set that cont...   
21   [What it is: ,  A eye mask set that will give ...   
25   [What it is:, \n,  An advanced, time-released,...   
26   [What it is: , A multitasking concealer that i...   
33   [What it is:  , A light, vegan, cooling eye ge...   
37   [What it is:, A cooling eye gel to minimize pu...   
38   [What it is: ,  An attachment head that is des...   
40   [What it is: , A cannabis-sativa-seed oil-infu...   
49   [What it is:  , A visibly lifting, firming, an...   
52   [What it is: , A silky, nourishing, coconut ey...   
91   [Which skin type is it good for?, ✔ Normal, ✔ ...   
106  [What it is:  , An oil-free, visibly brighteni...   
111  [Which skin type is it good for?, ✔ Normal, ✔ ...   
120  [What it is:  , A bestselling, all-in-one face...   
126  [What it is: ,  A bio-cellulose eye mask made ...   
127  [What it is:  , A weightless, ultra-hydrating,...   
128  [What 

In [60]:
moisturizers_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 645 entries, 0 to 644
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   brand              645 non-null    object
 1   name               645 non-null    object
 2   about the product  645 non-null    object
 3   weblink            645 non-null    object
 4   sub_category       645 non-null    object
 5   main_category      645 non-null    object
 6   num_likes          645 non-null    object
 7   img_link           645 non-null    object
 8   price              642 non-null    object
 9   size               569 non-null    object
 10  ingredients        645 non-null    object
 11  rating             645 non-null    object
 12  num_reviews        635 non-null    object
 13  highlights         645 non-null    object
 14  product_name       645 non-null    object
dtypes: object(15)
memory usage: 75.7+ KB


In [61]:
moisturizers_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 645 entries, 0 to 644
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   brand              645 non-null    object
 1   name               645 non-null    object
 2   about the product  645 non-null    object
 3   weblink            645 non-null    object
 4   sub_category       645 non-null    object
 5   main_category      645 non-null    object
 6   num_likes          645 non-null    object
 7   img_link           645 non-null    object
 8   price              642 non-null    object
 9   size               569 non-null    object
 10  ingredients        645 non-null    object
 11  rating             645 non-null    object
 12  num_reviews        635 non-null    object
 13  highlights         645 non-null    object
 14  product_name       645 non-null    object
dtypes: object(15)
memory usage: 75.7+ KB


In [62]:
moisturizers_raw[moisturizers_raw['size'].isnull()]

brand  \
8                 IT Cosmetics   
15                       Murad   
20                     Briogeo   
23                     Briogeo   
35                bareMinerals   
40         Dr. Brandt Skincare   
47                    CLINIQUE   
56                   Dr. Jart+   
81                        Dior   
119                      Murad   
124                   CLINIQUE   
128                   Caudalie   
130                      SK-II   
137               IT Cosmetics   
142                     La Mer   
147                  Biossance   
155               bareMinerals   
165                      SK-II   
167               IT Cosmetics   
168                    Skinfix   
174                    Vitruvi   
190                       LXMI   
200                      fresh   
206                   Caudalie   
207                MILK MAKEUP   
208                 Supergoop!   
216                       Dior   
225              Armani Beauty   
229                    Lancôme   
245                     La Mer   
257               AMOREPACIFIC   
265          Peter Thomas Roth   
267                   CLINIQUE   
276                   Shiseido   
281                   CLINIQUE   
284                Josie Maran   
291                   CLINIQUE   
296                   CLINIQUE   
298                   CLINIQUE   
302                      tarte   
314                      COOLA   
316                MILK MAKEUP   
321                   Smashbox   
352               Perricone MD   
355                      Murad   
359                    Lancôme   
360           First Aid Beauty   
373               IT Cosmetics   
376                      Murad   
379                      tarte   
396         REN Clean Skincare   
398               Estée Lauder   
399                  Dr. Jart+   
401                   CLINIQUE   
413                   CLINIQUE   
414                Bobbi Brown   
418                   CLINIQUE   
424                   CLINIQUE   
427                   CLINIQUE   
460         REN Clean Skincare   
476               Estée Lauder   
478               Perricone MD   
493               Estée Lauder   
512         SEPHORA COLLECTION   
524                Nurse Jamie   
526              Pretty Vulgar   
530                      fresh   
551                    Lancôme   
566                       LXMI   
571                   Erborian   
589              KORA Organics   
596                   Algenist   
598  Dr. Dennis Gross Skincare   
608          Peter Thomas Roth   
612         REN Clean Skincare   
637                   Caudalie   

                                                  name  \
8                  CC+ Cream Illumination with SPF 50+   
15                          Nutrient-Charged Water Gel   
20    B. Well Organic + Australian 100% Tea Tree Sk...   
23      B. Well Organic + Cold-Pressed 100% Castor Oil   
35   Prime Time™  BB Tinted Primer Broad Spectrum S...   
40                Hydro Biotic™ Recovery Sleeping Mask   
47                     Acne Solutions™ BB Cream SPF 40   
56                    Black Label Detox BB Beauty Balm   
81    Capture Totale Intensive Night Restorative Crème   
119  Essential-C Day Moisture Broad Spectrum SPF 30...   
124  Redness Solutions with Probiotic Technology Da...   
128                                      Beauty Elixir   
130     Mini Facial Treatment Essence (Pitera Essence)   
137               CC+ Cream Oil-Free Matte with SPF 40   
142           The Nourishing Soft Cream Hydration Duet   
147                      Squalane + Vitamin C Rose Oil   
155  COMPLEXION RESCUE™ Tinted Moisturizer with Hya...   
165          Facial Treatment Essence (Pitera Essence)   
167                             CC+ Cream with SPF 50+   
168           Barrier+ Triple Lipid-Peptide Face Cream   
174                             Sleep Face & Body Mist   
190           Crème du Nil Pore-Refining Moisture Veil   
200              Seaberry Skin Nutrition Cleansing Oil   
206           

Most of the rows with no associated 'size' are related to masks, supplements, devices, peels, make-up removers, gift sets, eye masks, eye gels, massager/rollers, etc. so these can be deleted as they are not the main focus of the study.

In [63]:
print('Initial shape: ', cleansers_raw.shape)
cleansers_raw.drop(cleansers_raw[cleansers_raw['size'].isnull()].index, inplace=True)
print('After deletion: ', cleansers_raw.shape)

Initial shape:  (409, 15)
After deletion:  (363, 15)


In [64]:
print('Initial shape: ', treatments_raw.shape)
treatments_raw.drop(treatments_raw[treatments_raw['size'].isnull()].index, inplace=True)
print('After deletion: ', treatments_raw.shape)

Initial shape:  (578, 15)
After deletion:  (498, 15)


In [65]:
print('Initial shape: ', eyeproducts_raw.shape)
eyeproducts_raw.drop(eyeproducts_raw[eyeproducts_raw['size'].isnull()].index, inplace=True)
print('After deletion: ', eyeproducts_raw.shape)

Initial shape:  (220, 15)
After deletion:  (190, 15)


In [66]:
print('Initial shape: ', moisturizers_raw.shape)
moisturizers_raw.drop(moisturizers_raw[moisturizers_raw['size'].isnull()].index, inplace=True)
print('After deletion: ', treatments_raw.shape)

Initial shape:  (645, 15)
After deletion:  (498, 15)


In [67]:
#Delete rows with no target variable - less than 10 per category
eyeproducts_raw.drop(eyeproducts_raw[eyeproducts_raw['price'].isnull()].index, inplace=True)
cleansers_raw.drop(cleansers_raw[cleansers_raw['price'].isnull()].index, inplace=True)
treatments_raw.drop(treatments_raw[treatments_raw['price'].isnull()].index, inplace=True)
moisturizers_raw.drop(moisturizers_raw[moisturizers_raw['price'].isnull()].index, inplace=True)

In [68]:
#Delete rows with no reviews 
eyeproducts_raw.drop(eyeproducts_raw[eyeproducts_raw['num_reviews'].isnull()].index, inplace=True)
cleansers_raw.drop(cleansers_raw[cleansers_raw['num_reviews'].isnull()].index, inplace=True)
treatments_raw.drop(treatments_raw[treatments_raw['num_reviews'].isnull()].index, inplace=True)
moisturizers_raw.drop(moisturizers_raw[moisturizers_raw['num_reviews'].isnull()].index, inplace=True)

Combining the dataframes before further cleaning...

In [69]:
df= cleansers_raw.append([treatments_raw, moisturizers_raw, eyeproducts_raw])

In [70]:
df.shape

(1586, 15)

**3) Removed 'stars' from rating and converted column into numeric type**

In [71]:
#Remove 'stars' from rating, convert to numeric
df['rating']= df.rating.str.partition(' ')[0]
df['rating']=pd.to_numeric(df['rating'])

**4) Remove currency sign from price and converted into numeric type**

In [72]:
#Remove $ from price, convert to numeric 
df['price']=df.price.str.strip('$')
df['price']=pd.to_numeric(df['price'])

**5) Translated both number of likes and number of reviews into digits and converted into numeric data**

In [73]:
#Cleaning num_likes column
df.loc[df.num_likes.str.endswith('K'), 'num_likes']=df.loc[df.num_likes.str.endswith('K'), 'num_likes'].str.strip('K').astype('float64')*1000
df['num_likes']=df['num_likes'].astype('int64')

In [74]:
#Cleaning num_reviews column
df.loc[df.num_reviews.str.endswith('K'), 'num_reviews']=df.loc[df.num_reviews.str.endswith('K'), 'num_reviews'].str.strip('K').astype('float64')*1000
df['num_reviews']=df['num_reviews'].astype('int64')

**6) Created new features based on skin type the product was compatible with. This was derived from the 'about the product section'.**

In [76]:
#Objective is to extract the skin type the product is compatible with from the 'about the product' section.
#Then create a new feature for each of the skin types

def search_in_list(term, list_desc):
    '''
    Returns 1 if any item on the list contains the term and 0 otherwise (term - lowercase)
    '''
    if any(term in i.lower() for i in list_desc):
        return 1
    else:
        return 0

In [77]:
#Create new features based on skintype
df['sensitive']= df['about the product'].apply(lambda row: search_in_list('sensitive', row))
df['combination']= df['about the product'].apply(lambda row: search_in_list('combination', row))
df['oily']= df['about the product'].apply(lambda row: search_in_list('oily', row))
df['normal']= df['about the product'].apply(lambda row: search_in_list('normal', row))
df['dry']= df['about the product'].apply(lambda row: search_in_list('dry', row))

In [78]:
#Inspect the output
df.head()

brand                                        name  \
0   Glow Recipe   Watermelon Glow PHA +BHA Pore-Tight Toner   
2        Tatcha         Pure One Step Camellia Oil Cleanser   
3          goop  GOOPGLOW Microderm Instant Glow Exfoliator   
4       Lancôme   Bi-Facil Double-Action Eye Makeup Remover   
5  OLEHENRIKSEN          Balancing Force™ Oil Control Toner   

                                   about the product  \
0  [What it is: , A bestselling gentle, PHA- and ...   
2  [Which skin type is it good for?, ✔ Normal, ✔ ...   
3  [What it is: ,  A clinically tested, dual-acti...   
4  [What it is:  , An award-winning makeup remove...   
5  [Which skin type is it good for?, \n, ✔ Normal...   

                                             weblink           sub_category  \
0  https://www.sephora.com/product/cleansing-exfo...                 Toners   
2  https://www.sephora.com/product/cleansing-exfo...  Face Wash & Cleansers   
3  https://www.sephora.com/product/cleansing-exfo...            Exfoliators   
4  https://www.sephora.com/product/cleansing-exfo...        Makeup Removers   
5  https://www.sephora.com/product/cleansing-exfo...                 Toners   

  main_category  num_likes                                           img_link  \
0     cleansers     125100  https://www.sephora.com/productimages/sku/s234...   
2     cleansers     107600  https://www.sephora.com/productimages/sku/s167...   
3     cleansers      12900  https://www.sephora.com/productimages/sku/s231...   
4     cleansers      58700  https://www.sephora.com/productimages/sku/s534...   
5     cleansers      92500  https://www.sephora.com/productimages/sku/s191...   

   price             size                                        ingredients  \
0   34.0  5.07 oz/ 150 mL  [ -Watermelon Extract: Hydrates, delivers esse...   
2   48.0   5.1 oz/ 150 mL  [ -Japanese Camellia Oil (Tsubaki): Seals in m...   
3  125.0    1.7 oz/ 50 mL  [ -Micro-exfoliating Minerals (Quartz, Garnet,...   
4   32.0   4.2 oz/ 125 mL  [Aqua/Water, Cyclopentasiloxane, Isohexadecane...   
5   29.0   6.5 oz/ 193 mL  [  -Green Fusion Complex™: Proprietary blend o...   

   rating  num_reviews                                         highlights  \
0     4.5         1900  [Good for: Pores , Good for: Dullness/Uneven T...   
2     4.5         1700  [Clean at Sephora, Hydrating, Best for Dry, Co...   
3     4.5         1200  [Clean at Sephora, Best for Dry, Combo, Normal...   
4     4.5         3700                                                 []   
5     4.5         1600  [Salicylic Acid, AHA/Glycolic Acid, Good for: ...   

                                        product_name  sensitive  combination  \
0  Glow RecipeWatermelon Glow PHA +BHA Pore-Tight...          0            1   
2          TatchaPure One Step Camellia Oil Cleanser          1            1   
3     goopGOOPGLOW Microderm Instant Glow Exfoliator          0            1   
4   LancômeBi-Facil Double-Action Eye Makeup Remover          1            1   
5     OLEHENRIKSENBalancing Force™ Oil Control Toner          1            1   

   oily  normal  dry  
0     1       1    1  
2     1       1    1  
3     1       1    1  
4     1       1    1  
5     1       1    1

**7) Created new feature based on highlights**

In [79]:
#First inspect what are the top categories for highlights 
highlights_list={}
for row in df['highlights']:
    for i in row:
        if i not in highlights_list.keys():
            highlights_list[i]=1 
        else:
            highlights_list[i]+=1

In [80]:
#highlights sorted by count
for item in sorted(highlights_list, key=highlights_list.get, reverse=True):
    print(item, highlights_list[item])

Clean at Sephora 577
Cruelty-Free 311
Good for: Dullness/Uneven Texture 291
Best for Dry, Combo, Normal Skin 285
Good for: Dryness 275
Good for: Anti-Aging 267
Vegan 234
Without Parabens  197
Best for Oily, Combo, Normal Skin 157
Good for: Loss of firmness 150
Hyaluronic Acid 146
Good for: Pores  134
Without Sulfates SLS & SLES 127
Community Favorite 121
Hydrating 106
Good for: Acne/Blemishes 105
Vitamin C 94
Recyclable Packaging 93
Fragrance Free 89
Salicylic Acid 77
AHA/Glycolic Acid 72
Without Phthalates  72
Good for: Dark Circles 71
Good for: Dark spots 68
Oil Free 56
Retinol 49
Niacinamide 45
Good for: Redness 44
Gluten Free 38
Alcohol Free 37
Best for Dry Skin 27
SPF 27
Hypoallergenic 25
Without Silicones 24
Best for Normal Skin 20
Lactic Acid 20
allure 2020 Best of Beauty Award Winner  19
Collagen 18
Best for Oily Skin 14
Without Mineral Oil  14
Sustainable Packaging 13
CBD 13
Plumping  11
Radiant Finish 9
Best for Combination Skin 8
allure 2018 Best of Beauty Award Winner 8
Ree

In [81]:
#highlights sorted alphabetically
for item in sorted(highlights_list):
    print(item, highlights_list[item])

AHA/Glycolic Acid 72
Alcohol Free 37
Best for Combination Skin 8
Best for Dry Skin 27
Best for Dry, Combo, Normal Skin 285
Best for Normal Skin 20
Best for Oily Skin 14
Best for Oily, Combo, Normal Skin 157
CBD 13
Clean at Sephora 577
Collagen 18
Community Favorite 121
Cream Formula 7
Cruelty-Free 311
Fragrance Free 89
Fresh Scent 2
Gluten Free 38
Good for: Acne/Blemishes 105
Good for: Anti-Aging 267
Good for: Dark Circles 71
Good for: Dark spots 68
Good for: Dryness 275
Good for: Dullness/Uneven Texture 291
Good for: Hair Dryness 2
Good for: Loss of firmness 150
Good for: Pores  134
Good for: Redness 44
High Shine Finish 1
Hyaluronic Acid 146
Hydrating 106
Hypoallergenic 25
Increases Shine 1
Lactic Acid 20
Liquid Formula 4
Long-wearing 1
Loose Powder Formula 1
Matte Finish 3
Natural Finish 1
Niacinamide 45
Oil Free 56
Plumping  11
Radiant Finish 9
Recyclable Packaging 93
Reef Safe SPF 8
Refill Available 3
Retinol 49
SPF 27
Salicylic Acid 77
Shimmer Finish 1
Stick Formula 1
Sustainable

Based on the results above, the several items have been merged under more general categories as the level of the detail is not necessary and this would also prevent creating too many features.  

1) Good for: 
- Acne/Blemishes
- Anti-Aging 
- Dark Circles  or Dark spots
- Dryness 
- Dullness/Uneven Texture 
- Loss of firmness
- Pores
- Redness
- Hydrating (to be added separately)
    
2) Without x / 'dirty' ingredients not used 
- Formaldehydes, Mineral Oil, Parabens, Phthalates, Retinyl Palmitate, Silicones
- Fragrance free (to be added separately)

3) Best for: 
- Combination, Dry, Combo, Normal, oily

4) Clean at Sephora 

5) Cruelty-Free 

6) Vegan

7) Skincare acids
- Hyaluronic Acid
- Salicylic Acid 
- AHA/Glycolic Acid
- Vitaminc C (to be added separately) 

8) Awards
- Community favorite
- any type of allure award

In [82]:
#Tackling 4,5,6 - more straight-forward 
df['clean']= df['highlights'].apply(lambda row: search_in_list('clean', row))
df['cruelty-free']= df['highlights'].apply(lambda row: search_in_list('cruelty', row))
df['vegan']= df['highlights'].apply(lambda row: search_in_list('vegan', row))

In [85]:
def extract_info2(phrase_add= None, phrase=None, list_=None):
    '''
    Function can:
    1) Search for term/phrase in the list and extracts the words immediately after the given phrase  
    2) Search for term/phrase and if found, this is added to the retrieved list  
    '''
    parsed=[]
    for highlight in list_:
        if phrase_add!=None:
            if phrase in highlight:
                parsed.append(highlight.partition(phrase)[2])
            if phrase_add in highlight:
                parsed.append(phrase_add)
        else:
            if phrase in highlight:
                parsed.append(highlight.partition(phrase)[2])    
    if not parsed:
        return None
    else:
        return parsed

In [87]:
#Using the custom function above, a similar process is undertaken for 1,2,3
df['skin_concerns']=df['highlights'].apply(lambda row: extract_info2('Hydrating', 'Good for: ', row))
df['excl_ingr']=df['highlights'].apply(lambda row: extract_info2('Fragrance', 'Without ', row))
df['best for']= df['highlights'].apply(lambda row: extract_info2(None, 'Best for ', row))

In [88]:
#Check output
df.iloc[11,13] #highlights
df.iloc[11, 23] #good for

['Alcohol Free',
 'Good for: Dullness/Uneven Texture',
 'Hydrating',
 'AHA/Glycolic Acid',
 'Clean at Sephora',
 'Community Favorite']

['Dullness/Uneven Texture', 'Hydrating']

In [89]:
#Check output
df.iloc[51, 13] #highlights
df.iloc[51, 24] #excluded ingredients

['Best for Oily, Combo, Normal Skin',
 'Good for: Acne/Blemishes',
 'Salicylic Acid',
 'Without Parabens ',
 'Without Sulfates SLS & SLES',
 'Fragrance Free']

['Parabens ', 'Sulfates SLS & SLES', 'Fragrance']

In [90]:
#Check output
df.iloc[534, 13] #highlights
df.iloc[534, 23] 
df.iloc[534, 24]
df.iloc[534, 25] 

'Snow Mushroom Water Serum'

['Best for Dry, Combo, Normal Skin',
 'Good for: Anti-Aging',
 'Hyaluronic Acid',
 'Cruelty-Free',
 'Vegan',
 'Clean at Sephora']

['Anti-Aging']

['Dry, Combo, Normal Skin']

In [91]:
#Inspect output
df.sample(12)

brand                                               name  \
576      Edible Beauty                   No. 2 Citrus Rhapsody Toner Mist   
13           BeautyBio  Bright Eyes Collagen-Infused Brightening Collo...   
115       SUNDAY RILEY             Pink Drink Firming Resurfacing Essence   
173      Edible Beauty                No. 4 Vanilla Silk Hydrating Lotion   
405          Sulwhasoo                         Bloomstay Vitalizing Cream   
415            Origins  Dr. Andrew Weil For Origins™ Mega-Mushroom Rel...   
270       AMOREPACIFIC       Treatment Enzyme Exfoliating Powder Cleanser   
41             Eve Lom                   WHITE Advanced Brightening Serum   
213   MAELYS Cosmetics                    B-PERKY Lift & Firm Breast Mask   
214           Shiseido                 White Lucent Brightening Gel Cream   
68            Algenist             ELEVATE Advanced Lift Contouring Cream   
177  Naturally Serious            Supercharge Anti-Oxidant Moisture Serum   

                                     about the product  \
576  [Which skin type is it good for?, ✔ Normal, ✔ ...   
13   [What it is: ,  A pair of eye masks that de-pu...   
115  [What it is:  , A peptide-infused essence that...   
173  [Which skin type is it good for?, ✔ Normal, ✔ ...   
405  [What it is: ,  A cushiony cream that helps vi...   
415  [Which skin type is it good for?, ✔ Normal, ✔ ...   
270  [What it is: , A powder to foam exfoliating da...   
41   [Which skin type is it good for?, \n, ✔ Normal...   
213  [What it is: ,  An ultra-revolutionary, breast...   
214  [What it is: ,  A refreshing, light, and smoot...   
68   [Which skin type is it good for?, ✔ Normal, ✔ ...   
177  [What it is: ,  A serum that helps protect fro...   

                                               weblink      sub_category  \
576  https://www.sephora.com/product/vitamin-c-sunc...  Mists & Essences   
13   https://www.sephora.com/product/skin-to-die-fo...         Eye Masks   
115  https://www.sephora.com/product/goodnight-glow...  Mists & Essences   
173  https://www.sephora.com/product/repairwear-las...      Moisturizers   
405  https://www.sephora.com/product/volition-beaut...      Moisturizers   
415  https://www.sephora.com/product/advanced-c-rad...            Toners   
270  https://www.sephora.com/product/dr-barbara-stu...      Facial Peels   
41   https://www.sephora.com/product/deep-sea-colla...       Face Serums   
213  https://www.sephora.com/product/dior-capture-t...        Face Masks   
214  https://www.sephora.com/product/dior-capture-t...      Moisturizers   
68   https://www.sephora.com/product/perfectionist-...      Moisturizers   
177  https://www.sephora.com/product/umbrian-clay-m...       Face Serums   

    main_category  num_likes  \
576  moisturizers       3600   
13   eye products       8900   
115  moisturizers       9500   
173  moisturizers       4600   
405  moisturizers       1700   
415  moisturizers       8600   
270    treatments      64000   
41     treatments       1400   
213  moisturizers       3600   
214  moisturizers       3400   
68     treatments       2700   
177    treatments       2400   

                                              img_link  price  \
576  https://www.sephora.com/productimages/sku/s205...   36.0   
13   https://www.sephora.com/productimages/sku/s216...   40.0   
115  https://www.sephora.com/productimages/sku/s237...   48.0   
173  https://www.sephora.com/productimages/sku/s205...   47.0   
405  https://www.sephora.com/productimages/sku/s229...  155.0   
415  https://www.sephora.com/productimages/sku/s210...   21.0   
270  https://www.sephora.com/productimages/sku/s218...   60.0   
41   https://www.sephora.com/productimages/sku/s185...  150.0   
213  https://www.sephora.com/productimages/sku/s226...   39.0   
214  https://www.sephora.com/productimages/sku/s221...   70.0   
68   https://www.sephora.com/productimages/sku/s195...   96.0   
177  https://www.sephora.com/productimages/sku/s212...   36.0   

        

Create new feature based on: 

- Skincare acids:
Hyaluronic Acid,
Salicylic Acid,
AHA/Glycolic Acid,
Vitamin C

- Awards:
Community favorite,
Allure winner

In [115]:
#Create a new feature for skincare acids used and any awards received - 7,8 
def find_and_encode(mapping, list_):
    '''
    Takes in a dictionary containing term to be searched and the string to be returned if term is found in the list
    '''
    word_list=[]
    for highlight in list_:
    #for each line 
        #look for each term 
        for term in mapping:
            if term in highlight:
                word_list.append(mapping[term])
            else:
                pass
    if not word_list:
        return None
    else:
        return word_list

In [116]:
#For skincare acids
term_map={'Hyaluronic': 'Hyaluronic Acid', 
'Salicylic':'Salicylic Acid',
'AHA':'AHA/Glycolic Acid',
'Vitamin C':'Vitamin C'}

In [117]:
df['acids']= df['highlights'].apply(lambda row: find_and_encode(term_map, row))

In [119]:
#Check output
df.iloc[12, 13] #highlights
df.iloc[12, 26]

['Clean at Sephora',
 'Community Favorite',
 'Salicylic Acid',
 'Good for: Acne/Blemishes',
 'Good for: Pores ',
 'Cruelty-Free']

In [121]:
#Check output
df.acids.value_counts()

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas/_libs/hashtable_class_helper.pxi", line 4588, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[Hyaluronic Acid]                       127
[Vitamin C]                              78
[Salicylic Acid]                         58
[AHA/Glycolic Acid]                      46
[AHA/Glycolic Acid, Salicylic Acid]      12
[Vitamin C, Hyaluronic Acid]              7
[AHA/Glycolic Acid, Hyaluronic Acid]      6
[Hyaluronic Acid, Vitamin C]              4
[Salicylic Acid, AHA/Glycolic Acid]       4
[Vitamin C, AHA/Glycolic Acid]            2
[Salicylic Acid, Vitamin C]               2
[Hyaluronic Acid, AHA/Glycolic Acid]      1
[Salicylic Acid, Hyaluronic Acid]         1
[AHA/Glycolic Acid, Vitamin C]            1
Name: acids, dtype: int64

In [122]:
#For awards
term_map2={'Community': 'Community favorite', 'allure':'allure winner'}

In [123]:
df['award']= df['highlights'].apply(lambda row: find_and_encode(term_map2, row))

In [124]:
#Check output
df.iloc[657, 13] #highlights
df.iloc[657, 27]

['Good for: Dullness/Uneven Texture',
 'Good for: Pores ',
 'Best for Dry, Combo, Normal Skin',
 'Lactic Acid',
 'Cruelty-Free',
 'Community Favorite']

['Community favorite']

In [126]:
#Check output
df.award.value_counts()

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas/_libs/hashtable_class_helper.pxi", line 4588, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[Community favorite]                   111
[allure winner]                         24
[allure winner, Community favorite]     10
Name: award, dtype: int64

**8) Create a new (uniform) feature for 'size'**

In [127]:
#Inspect size
df['size'].unique()

array(['5.07 oz/ 150 mL', '5.1 oz/ 150 mL', '1.7 oz/ 50 mL',
       '4.2 oz/ 125 mL', '6.5 oz/ 193 mL', '1.69 oz/ 50 mL',
       '5.41 oz/ 160 mL', '5 oz/ 150 mL', '1 oz/ 30 mL', '3.4 oz/ 100 mL',
       '6 oz/ 170.1 g', '6.7 oz/ 200 mL', '3.4 oz/100 mL', '8 oz/ 236 mL',
       '3.4 oz/ 101 mL', '4.0 oz/ 120 mL', '10.14 oz/ 300 mL',
       ' 2 oz / 67 mL', '20 Biodegradable Wipes', '2.7 oz/ 75 mL',
       '1 oz/ 29 mL', '3.3 oz/ 100 mL', 'Standard Size - 3.38 oz/ 100 mL',
       '6.76 oz/ 200 mL', '4.2 oz', '1 oz/ 28 g', '6 oz/ 180 mL',
       '3.2oz/100mL', '8.45 oz/ 250 mL', '12.85 oz/ 380 mL',
       '2 oz/ 59 mL', '5 oz/ 148 mL', '4 oz/ 120 ml', '4 oz/ 120 mL',
       '3.4 oz/ 115 ml', '0.68 oz/ 20 mL', '4.7 oz/ 125 mL',
       '5 oz/ 150 ml', '4 oz/120 mL', '8.4 oz/ 250 mL', '3.4 oz/ 100 g',
       '25 Wipes', '3 oz/ 88 g', '1.69 fl oz/ 50 mL', '3.2 oz/ 95 mL',
       '7.1 oz/ 201 g', '6.8 oz/ 200 mL', '28 x 0.01 oz/ 0.5 g packets',
       '4.05 oz', '3.5 oz/ 105 mL', '6.7 oz/ 198

Units used to note size are oz, ml, g.

From initial inspection, some entries have to be manually corrected. 

- 28 x 0.01 oz/ 0.5 g packets
- 10 x 2.27 oz cloths/ 10 x 67 mL cles oths
- 50 x 0.04 oz/ 1.25 mL
- 7 Ampoules, .0625 oz/ 1.8 mL each
- Four 0.27 oz/ 8 mL Vials
- 2 x 1.7 oz/ 50 mL
- 0.5 oz/ 15 mL x3
- 2 x 1 oz/ 30 mL
- 12 x 0.08 oz/ 2.2 g pre-saturated chemical peel pads
- 3 x 1.7 oz/ 50.28 mL
- 2 x 2.02 oz/ 60 mL

In [128]:
def return_match(pattern, source):
    '''
    Like a re.search function but fetches the exact match; none if no match 
    '''
    m=re.search(pattern, source)
    if m:
        return m.group()
    else:
        pass

In [129]:
#Create separate columns to capture all the data available - final unit to be decided on later
df['size_oz']=df['size'].apply(lambda x: return_match('[0-9. ]+o[Zz]', x))
df['size_ml']=df['size'].apply(lambda x: return_match('[0-9. ]+m[Ll]', x))
df['size_g']=df['size'].apply(lambda x: return_match('[0-9. ]+g', x))

In [131]:
#Inspect the values 
df.info()

1 oz         287
1.7 oz       249
0.5 oz       177
5 oz          52
2 oz          50
1.69 oz       43
6.7 oz        42
3.4 oz        37
4 oz          35
4.2 oz        29
1.6 oz        29
1.0 oz        28
5.07 oz       19
1.01 oz       19
6 oz          19
5.0 oz        16
5.1 oz        14
2.5 oz        14
3.3 oz        13
8.4 oz        12
 oz           12
8 oz          11
1.35 oz       11
2.53 oz       10
1.5 oz        10
6.76 oz        9
1.3 oz         9
8.5 oz         9
6.75 oz        9
3.38 oz        8
4.0 oz         7
6.8 oz         7
3 oz           7
2.6 oz         7
4.22 oz        6
0.85 oz        6
4.05 oz        6
0.67 oz        6
2.7 oz         5
1.02 oz        5
0.3 oz         5
1.18 oz        5
4.1 oz         5
0.68 oz        4
0.34 oz        4
0.35 oz        4
2.02 oz        3
6.5 oz         3
1.68 oz        3
5.41 oz        3
2.0 oz         3
1.8 oz         3
0.51 oz        3
3.6 oz         3
 1.7 oz        3
2.3 oz         3
0.95 oz        3
0.7 oz         3
0.33 oz       

Unit to be used --> oz

In [132]:
#Clean size_oz, removing trailing and leading whitespace and 'oz'
df['size_oz']=df['size_oz'].str.replace(' oz', '')
df['size_oz']=df['size_oz'].str.replace('oz', '')
df['size_oz']=df['size_oz'].str.strip()

In [133]:
#Convert into numeric type
df['size_oz']=pd.to_numeric(df['size_oz'])
df['size_oz'].dtype

Correct the errors spotted earlier by changing each entry individually  

In [136]:
#Manually correct the errors
df.loc[df['size'].str.contains('12 x 0.08 oz'), 'size_oz']=0.96
df.loc[df['size'].str.contains('10 x 2.27 oz'), 'size_oz']=22.7
df.loc[df['size'].str.contains('50 x 0.04 oz'), 'size_oz']=2
df.loc[df['size'].str.contains('Four 0.27 oz'), 'size_oz']=0.4375
df.loc[df['size'].str.contains('7 Ampoules'), 'size_oz']=0.4375
df.loc[df['size'].str.contains('Four 0.27 oz'), 'size_oz']=1.08
df.loc[df['size'].str.contains('2 x 1.7 oz'), 'size_oz']=1.08
df.loc[df['size'].str.contains('15 mL x3'), 'size_oz']=1.5
df.loc[df['size'].str.contains('2 x 1 oz'), 'size_oz']=2
df.loc[df['size'].str.contains('12 x 0.08 oz'), 'size_oz']=0.96
df.loc[df['size'].str.contains('3 x 1.7 oz'), 'size_oz']=5.1
df.loc[df['size'].str.contains('2 x 2.02 oz'), 'size_oz']=4.04

In [137]:
#Inspect the output
df[['size_ml', 'size_oz', 'size_g']]

size_ml  size_oz size_g
0     150 mL     5.07   None
2     150 mL     5.10   None
3      50 mL     1.70   None
4     125 mL     4.20   None
5     193 mL     6.50   None
..       ...      ...    ...
213    10 mL     0.34   None
214    15 mL     0.50   None
215    15 mL     0.50   None
216    15 mL     0.50   None
217    15 mL     0.50   None

[1586 rows x 3 columns]

In [138]:
#Inspect missing values for size and fill them out using other info where possible -- 67 missing entries
df[df['size_oz'].isnull()]

brand  \
31                       Kopari   
71           SEPHORA COLLECTION   
73           SEPHORA COLLECTION   
95           SEPHORA COLLECTION   
118          SEPHORA COLLECTION   
130                   BeautyBio   
136   Dr. Dennis Gross Skincare   
141                      NuFACE   
145                 Josie Maran   
147          SEPHORA COLLECTION   
172                OLEHENRIKSEN   
254                    Shiseido   
302          SEPHORA COLLECTION   
335  ONE/SIZE by Patrick Starrr   
363  ONE/SIZE by Patrick Starrr   
386                    CLINIQUE   
7              First Aid Beauty   
50           SEPHORA COLLECTION   
51              Volition Beauty   
115                   Peace Out   
144          SEPHORA COLLECTION   
167           Naturally Serious   
174                        goop   
188                     FARSÁLI   
221                       fresh   
227   Dr. Dennis Gross Skincare   
234           Peter Thomas Roth   
236          SEPHORA COLLECTION   
299          SEPHORA COLLECTION   
316                   Peace Out   
322                   Peace Out   
344   Dr. Dennis Gross Skincare   
353                   Peace Out   
356   Dr. Dennis Gross Skincare   
406        Dr. Zenovia Skincare   
427                        goop   
428   Dr. Dennis Gross Skincare   
437           Peter Thomas Roth   
467               Mario Badescu   
498          SEPHORA COLLECTION   
531   Dr. Dennis Gross Skincare   
538                     FARSÁLI   
555                      NuFACE   
567           Peter Thomas Roth   
570                     FARSÁLI   
38           SEPHORA COLLECTION   
227                        goop   
233                    Smashbox   
309                     Lancôme   
377          SEPHORA COLLECTION   
538                     Eve Lom   
606          Marc Jacobs Beauty   
4                   INC.redible   
5             Peter Thomas Roth   
13                    BeautyBio   
16                 Estée Lauder   
41                    Sulwhasoo   
55                 Ciaté London   
86                        tarte   
94                  MILK MAKEUP   
102  ONE/SIZE by Patrick Starrr   
112           Peter Thomas Roth   
118           Peter Thomas Roth   
122                     Clarins   
125                   Peace Out   
138                    Shiseido   
194           Peter Thomas Roth   

                                                  name  \
31                                  Coconut Melt Wipes   
71               Cleansing Wipes - Rose - Moisturizing   
73   Mini Triple Action Cleansing Water - Cleanse +...   
95                       Triple Action Cleansing Water   
118                     Waterprooof Eye Makeup Remover   
130  GloPRO® Prep Pads Clarifying Skin Cleansing Wi...   
136                          DRx Acne Eliminating Pads   
141                                Prep-N-Glow™ Cloths   
145                                   Bear Naked Wipes   
147                      Triple Action Cleansing Water   
172                Truth™ On the Glow Cleansing Cloths   
254                                      Facial Cotton   
302                Mini Waterprooof Eye Makeup Remover   
335                   Mini GO OFF Makeup Remover Wipes   
363                        GO OFF Makeup Remover Wipes   
386  Take The Day Off Micellar Cleansing Towelettes...   
7                                 Facial Radiance Pads   
50                              Hydrating Booster Pads   
51           Apple Cider Vinegar Resurfacing Peel Pads   
115              AHA + BHA + PHA Brightening Peel Pads   
144                               Clarifying Peel Pads   
167                    Get Even Cold-Pressed Peel Pads   
174  GOOPGLOW 5% Glycolic Acid Overnight Glow Peel ...   
188                        Liquid Glass Radiance Serum   
221              Rose Deep Hydration Oil-Infused Serum   
227  Alpha Beta® Ultra Gentle Daily Peel for Sensit...   
234       Peptide 21® Amino Acid Exfoliating Peel Pads   
236                  

In [139]:
#Inspecting each row
crit=df.size_oz.isnull()& df.size_ml.isnull()
df.loc[crit, ['size', 'size_oz', 'size_ml', 'size_g']]

size  size_oz size_ml size_g
31          20 Biodegradable Wipes      NaN    None   None
71                        25 Wipes      NaN    None   None
130                        30 pads      NaN    None   None
136                  45 Treatments      NaN    None   None
141  20 Individually Packed Cloths      NaN    None   None
145                       30 Wipes      NaN    None   None
172     30 facial cleansing cloths      NaN    None   None
254                      40 sheets      NaN    None   None
335           10 wipes -  7" x 10"      NaN    None   None
363            21 wipes - 7" x 10"      NaN    None   None
386                       50 Wipes      NaN    None   None
7                          28 Pads      NaN    None   None
50                         60 Pads      NaN    None   None
51                         30 Pads      NaN    None   None
115        8 Brightening Face Pads      NaN    None   None
144                        60 Pads      NaN    None   None
167                        60 pads      NaN    None   None
174                        12 Pack      NaN    None   None
227        30 treatments + 5 bonus      NaN    None   None
234                        60 Pads      NaN    None   None
236                        60 Pads      NaN    None   None
316                      6 patches      NaN    None   None
322          12 Microneedling Dots      NaN    None   None
344        30 Treatments + 5 Bonus      NaN    None   None
353           20 Acne Healing Dots      NaN    None   None
356       30 Treatments + 5 Bonus       NaN    None   None
406                        60 Pads      NaN    None   None
427                        12 Pack      NaN    None   None
428                  16 treatments      NaN    None   None
437                        60 pads      NaN    None   None
467                       Original      NaN    None   None
498                        20 Pads      NaN    None   None
531                  45 Treatments      NaN    None   None
555  20 Individually Packed Cloths      NaN    None   None
567                        72 dots      NaN    None   None
227                        12 Pack      NaN    None   None
233        Photo Finish Primerizer      NaN    None   None
309                   1 Sheet Mask      NaN    None   None
538                   1 Sheet Mask      NaN    None   None
4                         20 pairs      NaN    None   None
5             30 pairs/ 60 patches      NaN    None   None
13                        15 pairs      NaN    None   None
16                         4 pairs      NaN    None   None
55                 6 Patches x 3 g      NaN    None    3 g
86        4 x 2 single use patches      NaN    None   None
94                         6 pairs      NaN    None   None
102                      6 Patches      NaN    None   None
112    60 Pads-30 Treatments\t\t\t      NaN    None   None
118          30 pairs - 60 patches      NaN    None   None
125                       6x pairs      NaN    None   None
138         3 Packettes x 2 Sheets      NaN    None   None
194           30 Pairs/ 60 Patches      NaN    None   None

In [140]:
df.loc[df['size'].str.contains('6 Patches x 3 g'), 'size_oz']=0.106

In [141]:
#Other rows need to be dropped as there is no data available 
df.shape

(1586, 31)

In [142]:
df.drop(df[crit].index, inplace=True)

In [143]:
df.shape

(1426, 31)

In [144]:
#There may be entries for ml or g that we can use to compute oz
crit2=df.size_oz.isnull()& (df.size_ml.notnull() | df.size_g.notnull())
df.loc[crit2, ['size', 'size_oz', 'size_ml', 'size_g']]

size  size_oz size_ml size_g
73                1.69 fl oz/ 50 mL      NaN   50 mL   None
95               200ml / 6.76 fl oz      NaN   200ml   None
147              200ml / 6.76 fl oz      NaN   200ml   None
302               1.69 fl oz/ 50 mL      NaN   50 mL   None
188                   1 fl oz/ 30mL      NaN    30mL   None
221                          100 mL      NaN  100 mL   None
299   30 ml/ 1.01 fl oz\t\t\t\t\t\t      NaN   30 ml   None
570                  0.34fl oz/10mL      NaN    10mL   None
38                1.69 fl oz/ 50 mL      NaN   50 mL   None
377               1.69 fl oz/ 50 mL      NaN   50 mL   None
606                   3 fl oz/ 90ml      NaN    90ml   None
41                            20 mL      NaN   20 mL   None
122                           15 ml      NaN   15 ml   None

In [145]:
subset_ml=df.loc[crit2, 'size_ml']

In [146]:
#Clean the subset of size_ml required to calculate size_oz
subset_ml=subset_ml.str.lower().str.strip(' ml').astype('float64')

In [147]:
#Using 1ml = oz 0.033814
df.loc[crit2,'size_oz']=subset_ml.apply(lambda ml: ml*0.033814)

In [148]:
#Inspect the output
df.loc[crit2, ['size', 'size_oz', 'size_ml', 'size_g']]

size  size_oz size_ml size_g
73                1.69 fl oz/ 50 mL  1.69070   50 mL   None
95               200ml / 6.76 fl oz  6.76280   200ml   None
147              200ml / 6.76 fl oz  6.76280   200ml   None
302               1.69 fl oz/ 50 mL  1.69070   50 mL   None
188                   1 fl oz/ 30mL  1.01442    30mL   None
221                          100 mL  3.38140  100 mL   None
299   30 ml/ 1.01 fl oz\t\t\t\t\t\t  1.01442   30 ml   None
570                  0.34fl oz/10mL  0.33814    10mL   None
38                1.69 fl oz/ 50 mL  1.69070   50 mL   None
377               1.69 fl oz/ 50 mL  1.69070   50 mL   None
606                   3 fl oz/ 90ml  3.04326    90ml   None
41                            20 mL  0.67628   20 mL   None
122                           15 ml  0.50721   15 ml   None

**9) Derive a feature for price adjusted for the size or volume of the product**

In [149]:
#Create potential target variable (vs. price)
df['pricepervol']=df['price']/df['size_oz']

**10) Create new feature based on highlighted ingredients**

In [ ]:
#Inspect
df.iloc[92, 2]

In [151]:
#Extract the name of the ingredients highlighted in the 'about the product' section.
#Upon inspection, these ingredients follow a similar pattern so the approach will use regex

def find_in_list(regex, list_):
    '''
    Searches for pattern in the list and returns a list of the matches
    '''
    parsed=[]
    for line in list_:
        try:
            m=re.search(regex, line).group(0)
            parsed.append(m)
        except:
            pass
    
    if not parsed:
        return None
    else:
        return parsed

In [152]:
df['highlighted_ingr']=df['about the product'].apply(lambda x: find_in_list('(?<=- ).*?(?=:)', x))

**11) Create a new feature capturing whether clinical results are published in the about section**

In [153]:
df['clinical_results']=df['about the product'].apply(lambda x: search_in_list('results',x))

**12) Extract formulation type for each ingredient and capture this in a separate feature(s)**

In [154]:
#Creating custom function to capture type of formulation specified in 'about' section

def extract_next(phrase, list_):
    '''
    Searches for phrase in the list and extracts the next item on the list
    '''
    item_count=0
    for highlight in list_:
        item_count+=1
        if phrase in highlight:
            return list_[item_count]

In [155]:
#Testing if it works
extract_next('Formulation:', df.iloc[19, 2])

' Cream'

In [156]:
#Create new column extracting the formulation
df['formulation']=df['about the product'].apply(lambda x: extract_next('Formulation:', x))

In [157]:
#Inspect the output
df['formulation'].value_counts()

 Cream                                           75
Cream                                            64
Lightweight Serum                                63
Lightweight Liquid                               43
Lightweight Cream                                38
 Gel                                             31
Gel                                              28
Serum                                            27
Lightweight Serum\n                              22
 Serum                                           21
 Liquid                                          21
 Lightweight Serum                               21
Lightweight Gel                                  19
 Lightweight Cream                               19
Rich Cream                                       17
 Lightweight Gel                                 16
 Lightweight Serum                               15
 Lightweight Liquid                              15
Cream\n                                          14
Oil         

Output requires cleaning (for white spaces, \n, capitalisation) and consolidation. 
It seems that the data reveals the lightness or richness of the formula as well as the formulation (i.e. oil, serum, lotion, etc.) so two features will be created to capture these info after cleaning.

In [159]:
#Cleaning the formulation column
df['formulation']=df['formulation'].str.lstrip().str.rstrip('\n')

In [160]:
df['formulation']=df['formulation'].str.rstrip().str.lower()

In [162]:
df['formulation'].value_counts()

cream                                            167
lightweight serum                                138
lightweight liquid                                86
gel                                               77
lightweight cream                                 74
lightweight gel                                   54
serum                                             54
liquid                                            47
oil                                               41
rich cream                                        37
lightweight lotion                                31
scrub                                             20
lightweight oil                                   17
mask                                              14
lotion                                            12
rich serum                                         8
spray                                              6
rich lotion                                        5
lightweight mask                              

The following categories will be used:
- richness - light, normal, rich/heavy
- formulation - lotion, cream, serum, liquid, gel, oil, others

In [165]:
def find_and_encode2(mapping, default, row):
    '''
    Takes in a dictionary containing the terms to be searched and the value to return if term is found;
    if not found, the default is returned 
    '''
    x=default
    try:
        for i in mapping:
            if i in row:
                x=mapping[i]
                break
            else:
                pass
        return x
    except:
        return x

In [ ]:
#Creating the formulation column
formulation_map={'lotion': 'lotion', 'cream': 'cream', 'serum':'serum', 'liquid':'liquid', 'gel':'gel', 'oil':'oil'}

In [166]:
df['formulation_type']=df['formulation'].apply(lambda x: find_and_encode2(formulation_map, 'others', x))

In [167]:
df['formulation_type'].value_counts()

others    558
cream     288
serum     202
liquid    135
gel       132
oil        63
lotion     48
Name: formulation_type, dtype: int64

In [168]:
#Creating richness column
formula_map={'light': 'light', 'creamy': 'heavy', 'rich': 'heavy', 'heavy': 'heavy'}

In [169]:
df['richness']=df['formulation'].apply(lambda x: find_and_encode2(formula_map, 'normal', x))

### Check each column of interest

- product_name
- brand
- sub_category
- main_category
- rating
- num_reviews
- sensitive
- combination
- oily
- normal
- dry
- clean
- acids
- award
- cruelty-free
- vegan
- skin_concerns
- excl_ingr
- best for
- highlighted_ingr
- clinical_results
- formulation_type
- richness


- price
- pricepervol

In [171]:
df.sub_category.value_counts()

Face Serums                  337
Moisturizers                 311
Face Wash & Cleansers        190
Eye Creams & Treatments      153
Toners                        78
Face Oils                     71
Mists & Essences              69
Exfoliators                   33
Facial Peels                  32
Face Masks                    25
Blemish & Acne Treatments     24
Makeup Removers               17
Night Creams                  17
Mini Size                     10
Face Sunscreen                 8
Face Primer                    8
Setting Spray & Powder         7
Decollete & Neck Creams        7
Body Lotions & Body Oils       5
Moisturizer & Treatments       4
Eye Masks                      3
BB & CC Creams                 2
Face Wash                      2
For Face                       2
Skincare                       2
Makeup                         1
Scrub & Exfoliants             1
BB & CC Cream                  1
Eye Primer                     1
Eye Cream                      1
Highlighte

In [172]:
#Override subcategory for 'moisturizer and treatments'
df.loc[df.sub_category=='Moisturizer & Treatments',['sub_category']]=df.loc[df.sub_category=='Moisturizer & Treatments',['main_category']]

In [173]:
df['sub_category']=df.sub_category.str.lower()

In [174]:
df.iloc[170, 4]

'face wash & cleansers'

In [175]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1426 entries, 0 to 217
Data columns (total 37 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   brand              1426 non-null   object 
 1   name               1426 non-null   object 
 2   about the product  1426 non-null   object 
 3   weblink            1426 non-null   object 
 4   sub_category       1422 non-null   object 
 5   main_category      1426 non-null   object 
 6   num_likes          1426 non-null   int64  
 7   img_link           1426 non-null   object 
 8   price              1426 non-null   float64
 9   size               1426 non-null   object 
 10  ingredients        1426 non-null   object 
 11  rating             1426 non-null   float64
 12  num_reviews        1426 non-null   int64  
 13  highlights         1426 non-null   object 
 14  product_name       1426 non-null   object 
 15  sensitive          1426 non-null   int64  
 16  combination        1426 n

In [176]:
category_mapping={
    'exfoli': 'exfoliators and peels',
    'peels': 'exfoliators and peels',
    'eye': 'eye creams and treatments',
    'oils': 'essences, serums and treatments',
    'serum': 'essences, serums and treatments',
    'wash': 'face wash and cleansers',
    'essence': 'essences, serums and treatments',
    'masks': 'essences, serums and treatments',
    'blemish': 'essences, serums and treatments',
    'treatment': 'essences, serums and treatments',
    'creams': 'moisturizers and creams',
    'toners': 'toners',
    'moisturizers': 'moisturizers and creams',
    'cleansers': 'face wash and cleansers'
}

In [177]:
df['product_type']=df['sub_category'].apply(lambda x: find_and_encode2(category_mapping, 'others', x))

In [178]:
df['product_type'].value_counts()

essences, serums and treatments    532
moisturizers and creams            337
face wash and cleansers            193
eye creams and treatments          158
toners                              78
exfoliators and peels               66
others                              62
Name: product_type, dtype: int64

In [179]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1426 entries, 0 to 217
Data columns (total 38 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   brand              1426 non-null   object 
 1   name               1426 non-null   object 
 2   about the product  1426 non-null   object 
 3   weblink            1426 non-null   object 
 4   sub_category       1422 non-null   object 
 5   main_category      1426 non-null   object 
 6   num_likes          1426 non-null   int64  
 7   img_link           1426 non-null   object 
 8   price              1426 non-null   float64
 9   size               1426 non-null   object 
 10  ingredients        1426 non-null   object 
 11  rating             1426 non-null   float64
 12  num_reviews        1426 non-null   int64  
 13  highlights         1426 non-null   object 
 14  product_name       1426 non-null   object 
 15  sensitive          1426 non-null   int64  
 16  combination        1426 n

In [180]:
#Change others to the main category
df.loc[df.product_type=='others','product_type']=df.loc[df.product_type=='others','main_category']

In [181]:
df['product_type']=df['product_type'].apply(lambda x: find_and_encode2(category_mapping, 'others', x))

23  skin_concerns      748 non-null    object 
 24  excl_ingr          286 non-null    object 
 25  best for           457 non-null    object 
 26  acids              1426 non-null   object 
 27  award              1426 non-null   object 
 28  size_oz            1426 non-null   float64
 29  size_ml            1271 non-null   object 
 30  size_g             70 non-null     object 
 31  pricepervol        1426 non-null   float64
 32  highlighted_ingr   898 non-null    object 
 33  clinical_results   1426 non-null   int64  
 34  formulation        934 non-null    object 
 35  formulation_type   1426 non-null   object 
 36  richness           1426 non-null   object 

__________________________________________________________________________________________

## 2.0 The dataset to be used

In [185]:
df.columns

Index(['brand', 'name', 'about the product', 'weblink', 'sub_category',
       'main_category', 'num_likes', 'img_link', 'price', 'size',
       'ingredients', 'rating', 'num_reviews', 'highlights', 'product_name',
       'sensitive', 'combination', 'oily', 'normal', 'dry', 'clean',
       'cruelty-free', 'vegan', 'skin_concerns', 'excl_ingr', 'best for',
       'acids', 'award', 'size_oz', 'size_ml', 'size_g', 'pricepervol',
       'highlighted_ingr', 'clinical_results', 'formulation',
       'formulation_type', 'richness', 'product_type'],
      dtype='object')

In [186]:
columns_required= [
    'brand', 
    'product_name',
    'product_type',
    'num_likes', 
    'ingredients',
    'rating', 
    'num_reviews', 
    'sensitive', 
    'combination', 
    'oily', 
    'normal', 
    'dry', 
    'clean',
    'cruelty-free',
    'vegan', 
    'skin_concerns', 
    'excl_ingr', 
    'best for',
    'acids',
    'award', 
    'pricepervol',
    'highlighted_ingr',
    'clinical_results',
    'formulation_type', 
    'richness']

In [191]:
data=df[columns_required]

In [215]:
#Rename columns
new_colnames= [
    'brand', 
    'product_name',
    'product_type',
    'num_likes', 
    'ingredients',
    'rating', 
    'num_reviews', 
    'sensitive_type', 
    'combination_type', 
    'oily_type', 
    'normal_type', 
    'dry_type', 
    'clean_sephora',
    'cruelty_free',
    'vegan', 
    'skin_concerns', 
    'excl_ingr', 
    'best_for_skintype',
    'acids',
    'award', 
    'pricepervol',
    'highlighted_ingr',
    'clinical_results',
    'formulation_type', 
    'richness']

In [216]:
columns_map=dict(zip(columns_required, new_colnames))
data.rename(columns=columns_map, inplace=True)

In [203]:
data.shape

(1426, 26)

<font color='red'> BELOW STILL DRAFT </font> 

## 3.0 Inspecting and refining the selected dataset

Additional steps taken: 
- dropping duplicates
- addressing missing values 
- check distribution for any possible outliers

In [202]:
#Check for duplicates
duplicates= data.duplicated(subset=['product_name'], keep=False)
data[duplicates].sort_values('product_name')

index                      brand  \
258     338               AMOREPACIFIC   
1076    419               AMOREPACIFIC   
997     312                   Algenist   
748     569                   Algenist   
367      68                   Algenist   
1108    455                   Algenist   
231     297                   Algenist   
1059    393                   Algenist   
1321     77                   Algenist   
459     199                   Algenist   
1046    378                   Algenist   
439     175                   Algenist   
1005    324                  Biossance   
164     218                  Biossance   
306     398                   CLINIQUE   
471     215                   CLINIQUE   
167     222                   CLINIQUE   
415     147                   CLINIQUE   
644     432                   CLINIQUE   
171     228                   CLINIQUE   
267     349                   CLINIQUE   
711     520                   CLINIQUE   
284     371                   CLINIQUE   
326      15                   CLINIQUE   
48       64                   CLINIQUE   
392     104                   CLINIQUE   
308     400                   CLINIQUE   
179     241                   CLINIQUE   
899     188                   CLINIQUE   
715     524                   CLINIQUE   
799      60                   CLINIQUE   
311     403                   CLINIQUE   
325      12                   Caudalie   
771      24                   Caudalie   
62       84          Charlotte Tilbury   
1124    474          Charlotte Tilbury   
597     369                    Clarins   
1115    464                    Clarins   
923     219                Dermalogica   
440     176                Dermalogica   
821      84          Dr. Barbara Sturm   
507     260          Dr. Barbara Sturm   
1053    387        Dr. Brandt Skincare   
698     496        Dr. Brandt Skincare   
1032    361  Dr. Dennis Gross Skincare   
422     154  Dr. Dennis Gross Skincare   
241     310       Dr. Zenovia Skincare   
370      72       Dr. Zenovia Skincare   
245     317              Edible Beauty   
1210    576              Edible Beauty   
139     190               Estée Lauder   
843     110               Estée Lauder   
389      98                    FARSÁLI   
1231    603                    FARSÁLI   
573     336                 FENTY SKIN   
212     277                 FENTY SKIN   
86      121           First Aid Beauty   
404     126           First Aid Beauty   
867     150                Glow Recipe   
1377    158                Glow Recipe   
0         0                Glow Recipe   
756       0                Glow Recipe   
591     359                Glow Recipe   
1050    383                   Guerlain   
687     485                   Guerlain   
298     387                  Herbivore   
772      25                  Herbivore   
953     258                Josie Maran   
296     384                Josie Maran   
970     278            Kate Somerville   
726     540            Kate Somerville   
714     523            Kate Somerville   
1217    585            Kate Somerville   
638     423            Kate Somerville   
142     193            Kate Somerville   
555     314            Kate Somerville   
202     267            Kate Somerville   
505     258            Kate Somerville   
829      93            Kate Somerville   
1041    371         Kiehl's Since 1851   
433     166         Kiehl's Since 1851   
557     317         Kiehl's Since 1851   
859     134         Kiehl's Since 1851   
877     161         Kiehl's Since 1851   
860     135         Kiehl's Since 1851   
806      67                     Kopari   
314     407                     Kopari   
1075    417            LANCER Skincare   
382      87            LANCER Skincare   
365      66            LANCER Skincare   
1114    463            LANCER Skincare   
330      22            LANCER Skincare   
42       58            LANCER Skincare   
804      65                 Lord Jones   
493     243     

In [207]:
data=data.drop_duplicates(subset=['product_name'], keep='first')
data.shape

(1338, 26)

In [218]:
#Checking for missing values
data.isna().sum()

index                   0
brand                   0
product_name            0
product_type            0
num_likes               0
ingredients             0
rating                  0
num_reviews             0
sensitive_type          0
combination_type        0
oily_type               0
normal_type             0
dry_type                0
clean_sephora           0
cruelty_free            0
vegan                   0
skin_concerns         643
excl_ingr            1066
best_for_skintype     905
acids                1056
award                1215
pricepervol             0
highlighted_ingr      491
clinical_results        0
formulation_type        0
richness                0
dtype: int64

Missing values

- skin_concerns - TBD
- excl_ingr- TBD
- best_for_skintype - TBD
- acids               1056
- award               1215
- highlighted_ingr     491

In [212]:
#Imputing missing values
data.skin_concerns.value_counts()

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas/_libs/hashtable_class_helper.pxi", line 4588, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[Dryness]                                                                      74
[Dullness/Uneven Texture]                                                      49
[Anti-Aging]                                                                   42
[Acne/Blemishes]                                                               36
[Pores ]                                                                       35
[Hydrating]                                                                    24
[Anti-Aging, Dullness/Uneven Texture]                                          18
[Dryness, Dullness/Uneven Texture]                                             17
[Anti-Aging, Loss of firmness]                                                 17
[Loss of firmness]                                                             16
[Acne/Blemishes, Pores ]                                                       14
[Dark Circles]                                                                 12
[Dark spots]    

In [213]:
data.excl_ingr.value_counts()

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas/_libs/hashtable_class_helper.pxi", line 4588, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[Parabens ]                                                 60
[Fragrance]                                                 51
[Sulfates SLS & SLES]                                       24
[Sulfates SLS & SLES, Parabens , Phthalates ]               20
[Parabens , Sulfates SLS & SLES]                            18
[Parabens , Phthalates ]                                    16
[Silicones]                                                 15
[Sulfates SLS & SLES, Parabens ]                             7
[Fragrance, Parabens ]                                       6
[Sulfates SLS & SLES, Phthalates , Parabens ]                5
[Phthalates ]                                                4
[Mineral Oil ]                                               4
[Sulfates SLS & SLES, Mineral Oil ]                          4
[Parabens , Sulfates SLS & SLES, Phthalates ]                3
[Fragrance, Sulfates SLS & SLES]                             3
[Parabens , Phthalates , Sulfates SLS & SLES]          

In [219]:
data.best_for_skintype.value_counts()

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas/_libs/hashtable_class_helper.pxi", line 4588, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[Dry, Combo, Normal Skin]                              242
[Oily, Combo, Normal Skin]                             127
[Dry Skin]                                              23
[Normal Skin]                                           16
[Oily Skin]                                             12
[Combination Skin]                                       6
[Dry, Combo, Normal Skin, Oily, Combo, Normal Skin]      3
[Oily, Combo, Normal Skin, Dry, Combo, Normal Skin]      2
[Oily Skin, Combination Skin]                            2
Name: best_for_skintype, dtype: int64

In [220]:
data.acids.value_counts()

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas/_libs/hashtable_class_helper.pxi", line 4588, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[Hyaluronic Acid]                       102
[Vitamin C]                              68
[Salicylic Acid]                         45
[AHA/Glycolic Acid]                      38
[AHA/Glycolic Acid, Salicylic Acid]       7
[Vitamin C, Hyaluronic Acid]              5
[Hyaluronic Acid, Vitamin C]              4
[AHA/Glycolic Acid, Hyaluronic Acid]      3
[Salicylic Acid, AHA/Glycolic Acid]       3
[Vitamin C, AHA/Glycolic Acid]            2
[Salicylic Acid, Vitamin C]               2
[Hyaluronic Acid, AHA/Glycolic Acid]      1
[Salicylic Acid, Hyaluronic Acid]         1
[AHA/Glycolic Acid, Vitamin C]            1
Name: acids, dtype: int64

In [222]:
data.award.value_counts()

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas/_libs/hashtable_class_helper.pxi", line 4588, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[Community favorite]                   94
[allure winner]                        20
[allure winner, Community favorite]     9
Name: award, dtype: int64

In [223]:
data.highlighted_ingr.value_counts()

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas/_libs/hashtable_class_helper.pxi", line 4588, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[Jeju Green Tea Extract, Jeju Green Tea Seed Oil, Panthenol]                                                         3
[Compound K, Ginsenoside Re]                                                                                         3
[Reishi]                                                                                                             3
[Hydrogen Bio Water™, Dead Sea Salt, Coconut Water]                                                                  3
[Pitera™, InfinitPower Technology]                                                                                   3
                                                                                                                    ..
[Lactic Acid, Vitamin E, Chamomile Extract]                                                                          1
[Japanese Purple Rice, Okinawa Algae Blend and Hyaluronic Acid, Botanical Extracts]                                  1
[Salicylic Acid (BHA) 0.5%, Arginine, Provitamin

In [ ]:
#Check for any outliers etc

In [194]:
#Reset index
data.reset_index(inplace=True)

In [195]:
#Save as json or csv for next step 

RangeIndex(start=0, stop=1426, step=1)

In [1]:
df.product_type.value_counts()

NameError: name 'df' is not defined

In [193]:
profile=ProfileReport(data, title='Profiling Report', explorative=True)
profile.to_widgets()

Summarize dataset:   0%|          | 0/39 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render widgets:   0%|          | 0/1 [00:00<?, ?it/s]

## 4.0 Saving the dataset